### PR TITLE
Added support for replacing nested keys plist file,added shell script…

### PR DESCRIPTION
**Pull request by Javasamurai**
… for adding in build phase in resources,replacing it from manifest

### CODE REVIEW

File: tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj [hunk 1]
```diff
@@ -1427,11 +1416,13 @@
 				TargetAttributes = {
 					D5BC66E71635D0270059EC0F = {
 						CreatedOnToolsVersion = 9.0;
-						ProvisioningStyle = Manual;
+						DevelopmentTeam = CQ9DS8QV5K;
```
Comment (rampr) [inline tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj]: don't hardcode this

File: tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj [hunk 2]
```diff
@@ -1427,11 +1416,13 @@
...
+						ProvisioningStyle = Automatic;
 					};
 					E25AB2611F97557100069969 = {
 						CreatedOnToolsVersion = 9.0;
-						ProvisioningStyle = Manual;
+						DevelopmentTeam = CQ9DS8QV5K;
```
Comment (rampr) [inline tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj]: same as above, please don't hardcode

File: tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj [hunk 3]
```diff
@@ -1762,9 +1766,11 @@
 				CODE_SIGN_ENTITLEMENTS = TeaLeafIOS.entitlements;
 				CODE_SIGN_IDENTITY = "iPhone Developer";
 				"CODE_SIGN_IDENTITY[sdk=iphoneos*]" = "iPhone Developer";
+				CODE_SIGN_STYLE = Automatic;
 				COMPRESS_PNG_FILES = NO;
 				COPY_PHASE_STRIP = NO;
 				DEAD_CODE_STRIPPING = NO;
+				DEVELOPMENT_TEAM = CQ9DS8QV5K;
```
Comment (rampr) [inline tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj]: please remove hardcoding of team id

File: tealeaf/resources/extra [hunk 4]
```diff
@@ -0,0 +1 @@
+echo "extra script"
```
Comment (rampr) [inline tealeaf/resources/extra]: why is this needed ?

File: build.js [hunk 5]
```diff
@@ -242,6 +243,16 @@ function updateConfigPlist(app, config, plist) {
     });
 }
 
+function addExtrascript(app, path) {
```
Comment (rampr) [inline build.js]: what is this for ?

Comment (Javasamurai) [reply build.js]: Considering a situation where we have to run extra script in the build phase.It would take the content of shell from manifest and put in the **extra** shell script.

File: lib/updatePlist.js [hunk 6]
```diff
@@ -47,7 +46,22 @@ var getPlistConfigValue = function(plistConfig, manifest, key) {
...
   Object.keys(plistConfig).forEach(function(key) {
-    var value = getPlistConfigValue(plistConfig, manifest, key);
+    var value;
+
+    // For replacing keys in nested JSON values
+    if (typeof plistConfig[key] === "object" && !(plistConfig[key] instanceof Array)) {
+      Object.keys(plistConfig[key]).forEach(function (nested_key) {
+        value = getPlistConfigValue(plistConfig[key], manifest, nested_key);
+        if (value === undefined) {
+          return;
```
Comment (bijoshtj) [inline lib/updatePlist.js]: why do we need this ?

File: lib/updatePlist.js [hunk 7]
```diff
@@ -47,7 +46,22 @@ var getPlistConfigValue = function(plistConfig, manifest, key) {
...
+        }
+        obj.setVal(plistObj, key + "." + nested_key, value);
+        value = undefined;
```
Comment (bijoshtj) [inline lib/updatePlist.js]: why is this needed ?

File: tealeaf/resources/extra (hunk 4 above)
Comment (Javasamurai) [reply tealeaf/resources/extra]: For adding script in build phases in xcode project.

Comment (rampr) [reply tealeaf/resources/extra]: this shouldn't be needed anymore right ?

File: lib/updatePlist.js [hunk 8]
```diff
@@ -45,9 +44,40 @@ var getPlistConfigValue = function(plistConfig, manifest, key) {
   return value;
 };
 
+
+var replaceObjectRecursively = function (plistObj, manifest, plistConfig, key) {
+  var value;
+
+  Object.keys(plistConfig).forEach(function (nested_key) {
+    value = getPlistConfigValue(plistConfig, manifest, nested_key);
```
Comment (bijoshtj) [inline lib/updatePlist.js]: wont it break if plistConfig[key] is object ?

File: lib/updatePlist.js [hunk 9]
```diff
@@ -45,9 +44,40 @@ var getPlistConfigValue = function(plistConfig, manifest, key) {
...
+    if (value === undefined) {
+      return;
+    }
+    obj.setVal(plistObj, key.concat(nested_key).join("."), value);
+    if (typeof plistConfig[nested_key] === "object" && !(plistConfig[nested_key] instanceof Array)) {
```
Comment (bijoshtj) [inline lib/updatePlist.js]: cache plistConfig[nested_key] and use it

File: lib/updatePlist.js (hunk 8 above)
Comment (bijoshtj) [inline lib/updatePlist.js]:
```
var curr_val = plistConfig[nested_key];
if (curr_val is object) {
  replaceRecursive(plistObj, manifest, curr_val, key);
} else {
  value = getPlistConfigValue(plistConfig, manifest, nested_key);
  obj.setVal(plistObj, key.concat(nested_key).join("."), value);
}
```

wont this be enough?

Comment (Javasamurai) [reply lib/updatePlist.js]: No.It won't checked for values.

File: tealeaf/resources/extra (hunk 4 above)
Comment (bijoshtj) [inline tealeaf/resources/extra]: this wont be needed now.

Comment (Javasamurai) [reply tealeaf/resources/extra]: We need this script to replace.

Comment (bijoshtj) [reply tealeaf/resources/extra]: we are adding it from crashlytics. so this is unnecessarry

File: lib/updatePlist.js (hunk 8 above)
Comment (Javasamurai) [reply lib/updatePlist.js]: In that case Iterating won't be possible.

File: tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj [hunk 10]
```diff
@@ -1491,6 +1480,19 @@
...
+			buildActionMask = 2147483647;
+			files = (
+			);
+			inputPaths = (
+			);
+			outputPaths = (
+			);
+			runOnlyForDeploymentPostprocessing = 0;
+			shellPath = /bin/sh;
+			shellScript = ./resources/extra;
```
Comment (rampr) [inline tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj:1494]: these changes aren't needed right ?

File: tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj [hunk 11]
```diff
@@ -1373,7 +1362,6 @@
 			isa = PBXGroup;
 			children = (
 				E25AB2641F97557200069969 /* Stickers.xcassets */,
-				E25AB2661F97557200069969 /* Info.plist */,
```
Comment (rampr) [inline tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj:1376]: is this change okay ?

File: tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj (hunk 10 above)
Comment (Javasamurai) [reply tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj:1494]: No.These are definitely required as we need to add build phase to xcode project.

File: lib/updatePlist.js [hunk 12]
```diff
@@ -45,9 +44,42 @@ var getPlistConfigValue = function(plistConfig, manifest, key) {
...
 };
 
+
+var replaceObjectRecursively = function (plistObj, manifest, plistConfig, key) {
+  var value;
+
+  Object.keys(plistConfig).forEach(function (nested_key) {
+    var config_val = plistConfig[nested_key];
+
+    value = getPlistConfigValue(plistConfig, manifest, nested_key);
```
Comment (bijoshtj) [inline lib/updatePlist.js]: wont it break if plistConfig[nested_key] is an object?

File: lib/updatePlist.js [hunk 13]
```diff
@@ -45,9 +44,39 @@ var getPlistConfigValue = function(plistConfig, manifest, key) {
...
+
+    if (typeof config_val === 'object' && !(config_val instanceof Array)) {
+      key.push(nested_key);
+      replaceObjectRecursively(plistObj, manifest, plistConfig[nested_key], key);
+    } else {
+      value = getPlistConfigValue(plistConfig, manifest, nested_key);
+      obj.setVal(plistObj, key.concat(nested_key).join('.'), value);
+      return;
+    }
+    key.pop();
```
Comment (bijoshtj) [inline lib/updatePlist.js:62]: i think we can move this statement to above if block

Comment (Javasamurai) [reply lib/updatePlist.js:62]: Then the key value would be empty at first and we don't want it.

File: tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj (hunk 11 above)
Comment (Javasamurai) [reply tealeaf/TeaLeafIOS.xcodeproj/project.pbxproj:1376]: Xcode itself modified the file on opening.I am not sure whether we need it or not.

**Review by bijoshtj — APPROVED** [review]

**Comment by Javasamurai** [comment]
Moving these changes to new PR for hc and hc-sticker branch
